### PR TITLE
Add docker_container_exec module

### LINKED ISSUE #98: Feature Request: Exec functionality for docker_compose module
Currently, I do not see a way to interrace with containers leveraging the exec command.  It would be great if this was natively supported by this docker collection.

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #98.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
docker_container_exec

### CODE REVIEW

**Comment by felixfontein** [comment]
ready_for_review

File: plugins/modules/docker_container_exec.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/python
+#
+# Copyright 2016 Red Hat | Ansible
```
Comment (bmillemathias) [inline plugins/modules/docker_container_exec.py:3]: is it the standard header to set ?

Comment (felixfontein) [reply plugins/modules/docker_container_exec.py:3]: LOL, no, it's something I forgot to replace :D

File: plugins/module_utils/socket_handler.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
+# Copyright (c) 2019-2020, Felix Fontein <felix@fontein.de>
```
Comment (felixfontein) [inline plugins/module_utils/socket_handler.py:1]:
```suggestion
# Copyright (c) 2019-2021, Felix Fontein <felix@fontein.de>
```

File: plugins/module_utils/socket_helper.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# Copyright (c) 2019-2020, Felix Fontein <felix@fontein.de>
```
Comment (felixfontein) [inline plugins/module_utils/socket_helper.py:1]:
```suggestion
# Copyright (c) 2019-2021, Felix Fontein <felix@fontein.de>
```

File: plugins/modules/docker_container_exec.py (hunk 1 above)
Comment (felixfontein) [reply plugins/modules/docker_container_exec.py:3]:
```suggestion
# Copyright (c) 2021, Felix Fontein <felix@fontein.de>
```

Comment (felixfontein) [reply plugins/modules/docker_container_exec.py]: I've fixed it (and adjusted the dates in two other copyright lines).

**Review by bmillemathias — COMMENTED** [review]
There is also the README file to update to add the module to the list

**Comment by felixfontein** [comment]
@bmillemathias good point! There are also several other 'new' modules that never made it into README...

File: plugins/modules/docker_container_exec.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: docker_container_exec
+
+short_description: execute command in a docker container
```
Comment (Andersson007) [inline plugins/modules/docker_container_exec.py:14]:
```suggestion
short_description: Execute command in a docker container
```

File: plugins/modules/docker_container_exec.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  tty:
+    type: bool
+    default: false
+    description:
+      - Whether to allocate a TTY.
+
+extends_documentation_fragment:
+  - community.docker.docker
+  - community.docker.docker.docker_py_1_documentation
+
```
Comment (Andersson007) [inline plugins/modules/docker_container_exec.py:70]:
```suggestion
note:
  - Does not support C(check_mode).
```

File: plugins/modules/docker_container_exec.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            rc=result.get('ExitCode') or 0,
+        )
+    except NotFound:
+        client.fail('Could not find container "{0}"'.format(container))
+    except APIError as e:
+        if e.response and e.response.status_code == 409:
+            client.fail('The container "{0}" has been paused ({1})'.format(container, e))
+        client.fail('An unexpected docker error occurred: {0}'.format(e), exception=traceback.format_exc())
+    except DockerException as e:
+        client.fail('An unexpected docker error occurred: {0}'.format(e), exception=traceback.format_exc())
```
Comment (Andersson007) [inline plugins/modules/docker_container_exec.py:249]:
```suggestion
        client.fail('An unexpected docker error occurred: {0}'.format(e), exception=traceback.format_exc())
```
shouldn't we use `to_native()` here and everywhere?

**Review by Andersson007 — COMMENTED** [review]
Not being an expert in the field, left a couple of cosmetic things and I like the idea. @felixfontein thanks for the new module!

File: plugins/modules/docker_container_exec.py (hunk 6 above)
Comment (felixfontein) [reply plugins/modules/docker_container_exec.py:250]: Good question. All modules basically use the same error handling (without `to_native`), and so far I never heard about problems with it.

File: plugins/modules/docker_container_exec.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+  tty:
+    type: bool
+    default: false
+    description:
+      - Whether to allocate a TTY.
+
+extends_documentation_fragment:
+  - community.docker.docker
+  - community.docker.docker.docker_py_1_documentation
+note:
```
Comment (felixfontein) [inline plugins/modules/docker_container_exec.py:70]:
```suggestion
notes:
```

File: plugins/modules/docker_container_exec.py (hunk 6 above)
Comment (Andersson007) [reply plugins/modules/docker_container_exec.py:250]: I usually prefer to follow the doc when writing new stuff. IMO it's not crucial anyway, so as you like

Comment (felixfontein) [reply plugins/modules/docker_container_exec.py:250]: Since this is common to (almost) all modules, I think I'd prefer to fix it in all of them at once.

Comment (felixfontein) [reply plugins/modules/docker_container_exec.py:250]: (Also easier to undo for all at once if it turns out to cause problems somewhere...)

**Comment by felixfontein** [comment]
@bmillemathias @Andersson007 thanks a lot for reviewing!

**Pull request merged**